### PR TITLE
add feat: computational power

### PR DESCRIPTION
**Pull request by sakshee-19**
**Checklist**

- [X] My branch is up-to-date with the upstream `develop` branch.
- [ ] I have added necessary documentation (if appropriate).

**Which issue does this PR fix?**: 
fixes: #58 

**Why do we need this PR?**:
This would check if the ```query``` is a mathematical expression if so it would evaluate the expression and display the result. In mathematical expression it includes arithmetic operations ```+,-,/,*,%,^``` and ```trigonometric operations```. It would consider ```x``` and ```X``` as multiplication operator and ```^``` as power operator.

screenshot from 2018-02-27 02-03-50
screenshot from 2018-02-27 02-02-39

### CODE REVIEW

**Comment by sakshee-19** [comment]
@AdiChat @Pl4gue  Please review.

**Comment by Pl4gue** [comment]
Just reviewed it. Really nice, thank you :+1: 
However, float values aren't considered. Kindly add these! (`4.3*2` for example)
And maybe add `sqrt` (square root) as well.

**Comment by sakshee-19** [comment]
@Pl4gue Thanks.I will make the changes as soon as possible. :+1:

**Comment by sakshee-19** [comment]
@Pl4gue @AdiChat I have a doubt why it is showing conflict as I made the changes that way?

**Comment by sakshee-19** [comment]
@AdiChat @Pl4gue Please review.

**Comment by AdiChat** [comment]
Great work. :tada: 

In the results section, you could embed a calculator with current result tab as the display for it. This addition should be simple. A similar feature (that is a calculator) has been included in Quark. This will make this feature complete and will improve the user experience. 
You may, also, display the embedded calculator when specific keywords like `calculator`, `calculation` are searched for. 

Yes, you need to resolve the minor conflict. It came up due to formatting inconsistency (extra tab) from a previous merged pull request related to flake8. 

Kindly take a look at the above enhancement. :+1:

**Comment by sakshee-19** [comment]
@AdiChat I have a little doubt, I have seen the embedded calculator in ```quark```. Do you want to let the user enter ```Math expression``` in calculator only or you just want me to display the result in that calculator?

**Comment by AdiChat** [comment]
Ideally, a user should be able to search the expression and get the result as the current feature provides. :+1: 

The result should be displayed on the display for the embedded calculator. 
The idea is that after getting the result for an expression by searching, the user can go on to use the calculator for further calculations and this will improve the experience as it will save time and the user is likely to be familiar with a calculator. 

A difference from Quark is that in Quark, the calculator is a modal though, in Cosmos-Search, the calculator should be integrated into the result page and support additional operations that are currently supported in searching. :+1:

File: search/views.py [hunk 1]
```diff
@@ -37,6 +38,224 @@ def error404(request):
...
+    if op == 'sqrt':
+        return math.sqrt(val1)
+
+def check_precedence(opr, i,op_precdn, op_trigno):
+    if opr in op_precdn and i in op_trigno:
+        return True
+    if op_precdn.index(opr) <= 2:
+        return True
+    if op_precdn.index(opr) <= 5:
+        if op_precdn.index(i)>2:
```
Comment (arnavb) [inline search/views.py]: Minor refactoring:

```python
return op_precdn.index(i)>2
```

File: search/views.py [hunk 2]
```diff
@@ -37,6 +38,224 @@ def error404(request):
...
+            return True
+        else:
+            return False
+    else:
+        if op_precdn.index(i)>5:
```
Comment (arnavb) [inline search/views.py]: Here too.

File: search/views.py [hunk 3]
```diff
@@ -37,6 +38,224 @@ def error404(request):
...
+            else:
+                break
+    if value.size()==1 and operator.size()==0:
+        return value.peek()
+    else:
+        return False
+
+def getResult(expr):
+    op_precdn = ['u+','u-','^', '/', '*', '%', '+', '-']
+    op_trigno = ['sin', 'cos', 'cosec', 'tan', 'sec', 'cot', 'exp', 'sqrt']
```
Comment (arnavb) [inline search/views.py]: Calling this trigno is a bit misleading (considering trigno). Maybe just call it functions? Also, you might consider adding an option for specifying degrees/radians (using deg/rad as a unit for these trig functions)

File: search/views.py [hunk 4]
```diff
@@ -37,6 +38,224 @@ def error404(request):
...
+    tokens = isMathExpression(expr,op_trigno)
+    if tokens:
```
Comment (arnavb) [inline search/views.py]:
```Python
return evaluate(tokens,op_precdn,op_trigno) if tokens else False
```

File: search/views.py [hunk 5]
```diff
@@ -45,6 +264,13 @@ def query(request):
     data = json.loads(open(settings.METADATA_JSON, 'r').readline())
     ans = []
     rec = []
+    exprResult = None
+    res = getResult(q)
+    if type(res) == int or type(res) == float:
```
Comment (arnavb) [inline search/views.py]: Maybe ternary here.

**Comment by sakshee-19** [comment]
@AdiChat @Pl4gue Please review. I have separated the `css` file of `calculator.html` hence 'python manage.py collectstatic' command will be necessary here. Also, I know the `UI` of `calculator` is not the best, so please suggest I will change according to that. I have tested many expression and fixed the bugs I got. Let me know if any case fails in `computation`.

File: search/views.py [hunk 6]
```diff
@@ -121,7 +122,7 @@ def isMathExpression(expr, op_math):
                 j = j + 1
             try:
                 val = int(expr[i:j])
-            except:
+            except 'valueError':
```
Comment (arnavb) [inline search/views.py]: It should be `ValueError`

File: search/views.py [hunk 7]
```diff
@@ -232,14 +233,14 @@ def evaluate(tokens, op_precdn, op_math):
                         try:
                             val = value.pop()
                             value.push(calculate(val, -1, opr))
-                        except:
+                        except 'IndexError':
```
Comment (arnavb) [inline search/views.py]: Why is the exception in quotes?

**Comment by sakshee-19** [comment]
@AdiChat @Pl4gue Please review.

File: search/views.py [hunk 8]
```diff
@@ -74,65 +74,339 @@ def error500(request):
...
+            i = i + 3
+
+        else:
+            return False
+    return token
+
+
+def calculate(val1, val2, op):
+    if op == 'u-':
+        return - val1
```
Comment (arnavb) [inline search/views.py]: I would remove the space after the -. Otherwise it looks like binary minus.

File: search/views.py [hunk 9]
```diff
@@ -74,65 +74,339 @@ def error500(request):
...
+                return False
+            i = i + 5
+
+        elif expr[i:i + 4].lower() in op_math:
+            token.append(expr[i:i + 4].lower())
+            if i + 4 >= len(expr):
+                return False
+            elif expr[i + 4] != '(':
+                return False
+            i = i + 4
```
Comment (arnavb) [inline search/views.py]: You could use the `+=` operator for this, you know.

File: search/views.py [hunk 10]
```diff
@@ -74,65 +74,339 @@ def error500(request):
...
+    if op == 'exp':
+        return math.exp(val1)
+    if op == 'sqrt':
+        return math.sqrt(val1)
+    if op == 'log':
+        return math.sqrt(val1)
+
+
+def check_precedence(opr, i, op_precdn, op_math):
+    if opr in op_precdn and i in op_math:
```
Comment (arnavb) [inline search/views.py]: You could honestly shorten this function to one line using the logical operators, but that is a preference thing.

File: search/views.py [hunk 11]
```diff
@@ -74,65 +74,339 @@ def error500(request):
...
+                    val2 = value.pop()
+                    val1 = value.pop()
+                    operator.pop()
+                    value.push(calculate(val1, val2, opr))
+                except IndexError:
+                    return False
+            else:
+                break
+
+    if value.size() == 1 and operator.size() == 0:
```
Comment (arnavb) [inline search/views.py]:
```Python
return value.peek() if value.size() == 1 and operator.size() == 0 else False
```

File: search/views.py [hunk 12]
```diff
@@ -74,65 +74,339 @@ def error500(request):
...
+        return value.peek()
+    else:
+        return False
+
+
+def getResult(expr):
+    op_precdn = ['u+', 'u-', '^', '/', '*', '%', '+', '-']
+    op_math = ['sin', 'cos', 'cosec', 'tan', 'sec', 'cot', 'exp', 'sqrt', 'log']
+    tokens = isMathExpression(expr, op_math)
+    if tokens:
```
Comment (arnavb) [inline search/views.py]:
```Python
return evaluate(tokens, op_precdn, op_math) if tokens else False
```

File: search/views.py [hunk 13]
```diff
@@ -74,65 +74,339 @@ def error500(request):
...
+        return False
+
+
+# calculator
+def calculator(request):
+    q = request.POST.get('txt')
+    if q is not None:
+        getResult(q)
+        res = getResult(q)
+        if type(res) == int or type(res) == float:
```
Comment (arnavb) [inline search/views.py]: Ternary is possible here too.

**Comment by AdiChat** [comment]
Nice work :+1: 

From the UI point of view, you may use light colours (shades of grey, light blue, white) for the calculator. In the current state, the calculator seems a bit detached. You must place it under the search bar with a small gap. The expression should, also, be present along with the answer (maybe above the answer in smaller font). Your initial design was clean and nice. You may try to adapt the calculator to a lighter and smoother design. The design flow should be smooth across all features. 

From the point of functionality, it looks good. 

Instead of rendering `calculator.html` separately, you may include it in `searchresults.html` and make changes accordingly to the result number and other features. The idea is that as we move on to provide better results, for a particular expression, we can provide an insight which will be included below the calculator in the results section or a new div. Therefore, the calculator and searching should be integrated to provide a complete experience to the user. 

Kindly take a look into this :+1:

File: search/static/css/calculator.css [hunk 14]
```diff
@@ -5,25 +5,28 @@
...
 
 /* Inset shadow on the screen to create indent */
 .top .screen {
-    height: 75px;
+    height: 65px;
     width: 426px;
     float: left;
     padding: 10px 10px;
-    background: rgba(15, 5, 29, 0.5);
+    background: rgb(255, 255, 255);
```
Comment (arnavb) [inline search/static/css/calculator.css]: You could just write `white`... ;)

**Comment by sakshee-19** [comment]
@AdiChat @Pl4gue Please review.

File: search/static/css/calculator.css (hunk 14 above)
Comment (sakshee-19) [reply search/static/css/calculator.css]: Yes, but think this is also okay. :)

**Comment by AdiChat** [comment]
Nice work :+1: 

Currently, the calculator appears at the bottom for every search which should not happen. Instead of placing the calculator code in `searchresults.html`, you may keep the initial file `calculator.html` and use `{% include "calculator.html" %}`. Also, in the current state, calculator will be triggered when one clicks on calculator or enters an expression. As we go on to add more features, there may be cases when we have search results and also, a computed result. 

Additionally, take a look into enabling the user to use the calculator using keyboard. Also, the url for the calculator should be `http://search.opengenus.org/calculator/` instead of `http://search.opengenus.org/query/` to avoid conflicts with future apps. The current design looks good.

Kindly take a look into this :+1:

**Comment by sakshee-19** [comment]
@AdiChat  When there may be cases to have search results and a computed result as well for a `query`, should the calculator appear at the top followed by search results?

**Comment by AdiChat** [comment]
Yes, for cases with a computed result and search results, the calculator should appear at the top. 

For the case of clicking on "calculator", you may render it directly from `calculator.html`.

**Comment by sakshee-19** [comment]
Okay thanks I will make the changes soon. :)

**Comment by AdiChat** [comment]
The design and functionality looks good. :+1: 

You should move the functions related to calculator from `views.py` to a separate file `calculator.py`. This is to keep the code contributor friendly. 

When tying in the calculator, pressing `enter` key should trigger evaluation `=`. Also, when we browse to `http://search.opengenus.org/calculator/`, the cursor should be placed within the calculator to facilate immediate typing. 

Kindly take a look into this. :+1:

**Comment by sakshee-19** [comment]
@AdiChat Please review.

**Comment by AdiChat** [comment]
@iattempt Template tags is a convention to store code to process any data and add it to any template irrespective of the view. It works. :+1:

**Comment by AdiChat** [comment]
Looks perfect. :+1: 

When we type an expression, the part `Showing result for expression: [icons]` appears which needs to be modified. You may place the result of computation there like `Showing result for expression: 1+1 = 2 [icons])`. You may, also, try out other possibilities. Also, you may use `Error` instead of `NaN`. 

Kindly take a look into this minor change. :+1:

**Comment by sakshee-19** [comment]
@AdiChat Intitially I have done that. But, then I thought It would be redundant as we are showing the same thing in display of calculator.

**Comment by AdiChat** [comment]
I guess it will highlight the result and will drive attention to the share buttons.

**Comment by sakshee-19** [comment]
@AdiChat Its done :+1: Please Review.

**Comment by AdiChat** [comment]
Looks perfect 😍 

Hope you are enjoying your journey 🚆  with **OpenGenus**

**Pull request merged**